src/averageAges.js

```diff
@@ -1,61 +1,53 @@
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const menDied = century
+    ? people.filter(person => person.sex === 'm'
+    && Math.ceil(person.died / 100) === century)
+    : people.filter(person => person.sex === 'm');
+  const menAge = menDied.map(person => person.died - person.born);
+  const menAverageAge = menAge.reduce((sum, age) => {
+    return sum + age;
+  }, 0) / menAge.length;
+
+  return menAverageAge;
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter(person => people.some(child => {
+      return person.name === child.mother;
+    })
+    && person.sex === 'f')
+    : people.filter(person => person.sex === 'f');
+  const womenAge = women.map(person => person.died - person.born);
+  const womenAverageAge = womenAge.reduce((sum, age) => {
+    return sum + age;
+  }, 0) / womenAge.length;
+
+  return womenAverageAge;
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = onlyWithSon
+    ? people.filter(person => people.some(child => {
+      return child.name === person.mother;
+    })
+    && person.sex === 'm')
+    : people.filter(person => people.some(child => {
+      return child.name === person.mother;
+    }));
+  const ageDiff = mothers.map((child) => {
+    const mother = people.find(person => child.mother === person.name);
+    const ages = child.born - mother.born;
+
+    return ages;
+  });
+  const AverageAgeDiff = ageDiff.reduce((sum, age) => {
+    return sum + age;
+  }, 0) / ageDiff.length;
+
+  return AverageAgeDiff;
```

Comment:
there is no need to create separate variable
```suggestion
  return ageDiff.reduce((sum, age) => {
    return sum + age;
  }, 0) / ageDiff.length;
```